File: JET/modul_2_oop/L02Datenkapselung/Exercise2-state_driven_hamster/Hamster.cs
namespace Hamster;

public class Hamster
{
    // Felder
    private static string _hungryRepresentation = "😡";
    private static string _fedRepresentation = "🐹";

    // Eigenschaften (Properties)
    public (int x, int y) Position { get; set; }
    public string Representation { get; private set; }

    public bool IsHungry { get; private set; }

    // Beziehungen
    private Plane plane;
    private List<Seedling> mouth = new List<Seedling>();

    // Konstruktor
    public Hamster(Plane plane)
    {
        IsHungry = false;
        Representation = _fedRepresentation;
        this.plane = plane;

        PositionAndManageHamster();
    }

    // Methoden
    private void PositionAndManageHamster()
    {
        var random = new Random();
        bool done;
        int x, y;

        do
        {
            x = random.Next(plane.Size);
            y = random.Next(plane.Size);

            done = plane.AssignInitialPosition(this, (x, y));
        } while (!done);

        Position = (x, y);
    }

    public void Move()
    {
        var random = new Random();
        int index = random.Next(Enum.GetValues<Direction>().Length);
        var direction = Enum.GetValues<Direction>()[index];

        plane.Position(this, direction);
    }

    public void NutritionBehaviour()
    {
        var random = new Random();

        // Zufällig hungrig werden
        if (random.NextDouble() < 0.1)
        {
            IsHungry = true;
            Representation = _hungryRepresentation;
        }

        if (plane.ContainsSeedling(Position))
        {
            if (IsHungry)
            {
                EatSeedlingFromTile();
            }
            else
            {
                PutInMouth();
            }
        }
        else
        {
            if (IsHungry && mouth.Any())
            {
                EatSeedlingFromMouth();
            }
        }
    }

    private void EatSeedlingFromMouth()
    {
        Eat();
        mouth.RemoveAt(0);
    }

    public void EatSeedlingFromTile()
    {
        Eat();
        plane.HamsterIsEatingSeedlings(this);
    }

    private void Eat()
    {
        IsHungry = false;
        Representation = _fedRepresentation;
    }

    public void PutInMouth()
    {
        var samen = plane.GetSeedlingOn(Position);
        mouth.Add(samen);
        plane.HamsterIsStoringSeedlings(this);
    }

    public override string ToString()
    {
        return Representation;
    }
}

File: BITEC/modul_2_object_orientation/src/lerneinheiten/L02KlassenUndMethoden/hamster_mit_data_hiding/Hamster.java
package lerneinheiten.L02KlassenUndMethoden.hamster_mit_data_hiding;

import java.util.ArrayList;
import java.util.Random;

public class Hamster {
    // Felder
    private int xPosition;
    private int yPosition;
    private String _representation;
    private final static String _hungryRepresentation = "🤬";
    private final static String _fedRepresentation = "🐹";
    private boolean _isHungry;
    private ArrayList<Seedling> _mouth = new ArrayList<>();
    private final Random _random = new Random();

    // Hat-Beziehungen
    Plane _plane;
    // -----------------------------
    // Methoden
    public void nahrungsVerhalten() {
        // Zufällig hungrig werden
        if (_random.nextDouble() < 0.1)
        {
            _isHungry = true;
            _representation = _hungryRepresentation;
        }

        boolean tileHasSeedling = _plane.tileTakenBySeedling(xPosition, yPosition);

        if (tileHasSeedling)
        {
            if (_isHungry)
            {
                eatSeedlingFromTile();
            }
            else
            {
                storeInMouth();
            }
        }
        else
        {
            if (_isHungry && !_mouth.isEmpty())
            {
                eatSeedlingFromMouth();
            }
        }
    }

    private void eat() {
        _representation = _fedRepresentation;
        _isHungry = false;
    }

    private void eatSeedlingFromMouth()
    {
        eat();
        _mouth.remove(0);
    }

    private void eatSeedlingFromTile()
    {
        eat();
        _plane.hamsterIsEatingSeedlings(this);
    }

    private void storeInMouth()
    {
        Seedling seedling = _plane.getSeedlingOn(xPosition, yPosition);
        _mouth.add(seedling);
        _plane.hamsterIsStoringSeedlings(this);
    }

    void bewegen() {
        int index = _random.nextInt(0, Direction.values().length);
        Direction direction = Direction.values()[index];

        _plane.bewegeHamster(this, direction);
    }

    // Konstruktor
    public Hamster(Plane plane) {
        boolean done;
        _plane = plane;
        _representation = _hungryRepresentation;
        _isHungry = true;
        int xWunsch;
        int yWunsch;

        // zufällig den Hamster positionieren.
        do {
            Random random = new Random();
            xWunsch = random.nextInt(0, _plane.getSize());
            yWunsch = random.nextInt(0, _plane.getSize());

            done = _plane.assignInitialPosition(this, xWunsch, yWunsch);
        } while (!done);

        xPosition = xWunsch;
        yPosition = yWunsch;
    }

    // Get- und Set-Methoden
    public int getxPosition() {
        return xPosition;
    }

    public void setxPosition(int xPosition) {
        this.xPosition = xPosition;
    }

    public int getyPosition() {
        return yPosition;
    }

    public void setyPosition(int yPosition) {
        this.yPosition = yPosition;
    }

    public String getRepresentation() {
        return _representation;
    }

    public void setRepresentation(String _representation) {
        this._representation = _representation;
    }

    public boolean getIsHungry() {
        return _isHungry;
    }

    public void setIsHungry(boolean _isHungry) {
        this._isHungry = _isHungry;
    }

    public ArrayList<Seedling> getMouth() {
        return _mouth;
    }

    public void setMouth(ArrayList<Seedling> _mouth) {
        this._mouth = _mouth;
    }

    public Plane getPlane() {
        return _plane;
    }

    public void setPlane(Plane _plane) {
        this._plane = _plane;
    }
}
